src/cljam/io/vcf/util/validator.clj

```diff
@@ -0,0 +1,335 @@
+(ns cljam.io.vcf.util.validator
+  (:require [cljam.io.vcf.util :as vcf.util]))
+
+(defrecord VCFValidator [ploidy file-type defs samples])
+
+(defn- error [path msg & args]
+  {:errors {path [(apply format msg args)]}})
+
+(defn- warn [path msg & args]
+  {:warnings {path [(apply format msg args)]}})
+
+(defn- merge-validation-results
+  ([] nil)
+  ([res] res)
+  ([res1 res2]
+   (letfn [(rec [x y]
+             (cond (nil? x) y
+                   (nil? y) x
+                   (map? x) (merge-with rec x y)
+                   :else (into x y)))]
+     (rec res1 res2)))
+  ([res1 res2 res3 & more]
+   (reduce merge-validation-results res1 (list* res2 res3 more))))
+
+(defn- prepend-keys [res keys]
+  ;; takes {:errors {[:chr] ["foo"]}}
+  ;; returns {:errors {[key ... :chr] ["foo"]}}
+  (when-not (nil? res)
+    (into {}
+          (map (fn [[kind m]]
+                 [kind
+                  (into {} (map (fn [[path messages]]
+                                  [(into keys path) messages]))
+                        m)]))
+          res)))
+
+(defn- error-on-bcf [validator path & args]
+  (let [f (if (= :bcf (:file-type validator)) error warn)]
+    (apply f path args)))
+
+(defn- prep-contig-defs [_validator contigs]
+  (into {} (map (juxt :id identity)) contigs))
+
+(defn- validate-contig [validator {:keys [chr]}]
+  (if (and (string? chr) (re-matches #"[^<>, \t]+" chr))
```

Comment:
According to [VCFv4.4 Specification](https://github.com/samtools/hts-specs/blob/9ea3a9d1fe1f18579f9a429ef224a8d05dbc48f1/VCFv4.4.tex#L335), I thought it was permissible to use `<`, `>`.

Reply:
Thanks for pointing it out! Honestly, I didn't have much time when I was writing the code, so I just borrowed that condition from the original code.

But, after reconsulting the spec, I found the following descriptions:

> It is recommended for VCF, and required for BCF, that the header includes tags describing the contigs referred to in the file.
> https://github.com/samtools/hts-specs/blob/9ea3a9d1fe1f18579f9a429ef224a8d05dbc48f1/VCFv4.4.tex#L246

> Contig names follow the same rules as the SAM format’s reference sequence names: they may contain any printable ASCII characters in the range [!-\~] apart from ‘\ , "‘’ () [] {} <>’ and may not start with ‘\*’ or ‘=’. Thus they match the following regular expression:
>   [0-9A-Za-z!#$%&+./:;?@^\_|\~-][0-9A-Za-z!#$%&\*+./:;=?@^\_|\~-]\*
> https://github.com/samtools/hts-specs/blob/9ea3a9d1fe1f18579f9a429ef224a8d05dbc48f1/VCFv4.4.tex#L266-L271

From these descriptions alone, it’s not clear whether CHROM should follow the naming rule of contigs. However, you can also find the following statement in the spec:

> The CHROM field in BCF2 is encoded as an integer offset into the list of ##contig field headers in the VCF header.
> https://github.com/samtools/hts-specs/blob/9ea3a9d1fe1f18579f9a429ef224a8d05dbc48f1/VCFv4.4.tex#L1975

With this in mind, at least in BCF, CHROM must be a contig defined in the header, which means CHROMs need to obey contig’s naming rule. Ideally, I think it’s the case with VCF as well.

So, to wrap up, I think the regex here should be:

```
[0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*
```

If you agree with this interpretation, I'll fix the code as such.

Reply:
Thank you for your suggestion. 
I think it's reasonable, so I agree to change the code accordingly 👍